Property-based tests for encoder and decoder

How this test should looks like:

1. Create big data type with all possible types supported by `tomland`.
2. Write converter for this type.
3. Ensure `decode . encode == id`.

(moved from separate issue to here)

I'm getting the weirdest issue. Just adding a module `Test.Toml.Codec.Property` and importing Hedgehog will break something in the `Spec.hs` module.

Basically if I run `cabal new-test` or `stack test`, the whole test-suite runs fine, but if I try to open it in a repl with `cabal new-repl tomland:tomland-test` or the stack equivalent, then I get

```
test/Spec.hs:7:1: error:
    Could not find module ‘Test.Tasty.QuickCheck’
```

Then, if I add `tasty-quickcheck ^>= 0.10` to tomland.cabal, then `cabal new-test` fails with the following:

```
test/Spec.hs:38:16: error:
    • No instance for (QC.Testable Hedgehog.Internal.Property.Property)
        arising from a use of ‘QC.testProperty’
      There are instances for similar types:
        instance [safe] QC.Testable QC.Property
          -- Defined in ‘QuickCheck-2.12.6.1:Test.QuickCheck.Property’
    • In the second argument of ‘($)’, namely
        ‘QC.testProperty "reverse" Test.Toml.Codec.Property.prop_reverse’
      In a stmt of a 'do' block:
        t2 <- pure
                $ QC.testProperty "reverse" Test.Toml.Codec.Property.prop_reverse
      In the expression:
        do t0 <- testGroup
                   "tomlRoundtrip" Test.Toml.Parsing.Property.test_tomlRoundtrip
           t1 <- HS.testSpec "Parser" Test.Toml.Parsing.Unit.spec_Parser
           t2 <- pure
                   $ QC.testProperty "reverse" Test.Toml.Codec.Property.prop_reverse
           t3 <- testGroup
                   "PrefixTreeInsertLookup"
                   Test.Toml.PrefixTree.Property.test_PrefixTreeInsertLookup
           ....
```

--- a/test/Test/Toml/BiCode/Property.hs
+++ b/test/Test/Toml/BiCode/Property.hs
@@ -3,7 +3,6 @@ module Test.Toml.BiCode.Property where
 import Control.Applicative ((<|>))
 import Control.Category ((>>>))
 import Data.ByteString (ByteString)
-import Data.Hashable (Hashable)
 import Data.HashSet (HashSet)
 import Data.IntSet (IntSet)
 import Data.List.NonEmpty (NonEmpty)
@@ -17,11 +16,11 @@ import Numeric.Natural (Natural)
 import Toml (TomlBiMap, TomlCodec, (.=))
 import Toml.Bi.Code (decode, encode)
 
-import Test.Toml.Gen (PropertyTest, genBool, genDay, genDouble, genHours, genInteger, genLocal,
-                      genText, genZoned, prop)
+import Test.Toml.Gen (PropertyTest, genBool, genByteString, genDay, genDouble, genFloat, genHashSet,
+                      genHours, genInt, genIntSet, genInteger, genLByteString, genLocal, genNatural,
+                      genNonEmpty, genString, genText, genWord, genZoned, prop)
 
 import qualified Data.ByteString.Lazy as L
-import qualified Data.Text as T
 import qualified Hedgehog.Gen as Gen
 import qualified Hedgehog.Range as Range
 import qualified Toml
@@ -144,8 +143,8 @@ genBigType = do
     btFloat         <- Batman <$> genFloat
     btText          <- genText
     btString        <- genString
-    btBS            <- genBS
-    btLazyBS        <- L.fromStrict <$> genBS
+    btBS            <- genByteString
+    btLazyBS        <- genLByteString
     btLocalTime     <- genLocal
     btDay           <- genDay
     btTimeOfDay     <- genHours
@@ -154,39 +153,14 @@ genBigType = do
     btArrayIntSet   <- genIntSet
     btArrayHashSet  <- genHashSet genNatural
     btArrayNonEmpty <- genNonEmpty genText
-    btNonEmpty      <- genNonEmpty genBS
+    btNonEmpty      <- genNonEmpty genByteString
     btList          <- Gen.list (Range.constant 0 5) genBool
     btNewtype       <- genNewType
     btSum           <- genSum
     btRecord        <- genRec
     pure BigType {..}
 
-genInt :: Gen Int
-genInt = Gen.int $ Range.constant 0 256
-
-genNatural :: Gen Natural
-genNatural = fromIntegral <$> Gen.int (Range.constant 1 256)
-
-genWord :: Gen Word
-genWord = Gen.word $ Range.constant 0 256
-
-genFloat :: Gen Float
-genFloat = Gen.float (Range.constant 0 256)
-
-genString :: Gen String
-genString = T.unpack <$> genText
-
-genBS :: Gen ByteString
-genBS = Gen.utf8 (Range.constant 0 20) Gen.alphaNum
-
-genIntSet :: Gen IntSet
-genIntSet = fromList <$> Gen.list (Range.constant 0 5) genInt
-
-genHashSet :: (Eq a, Hashable a) => Gen a -> Gen (HashSet a)
-genHashSet genA = fromList <$> Gen.list (Range.constant 0 5) genA
-
-genNonEmpty :: Gen a -> Gen (NonEmpty a)
-genNonEmpty = Gen.nonEmpty (Range.constant 1 5)
+-- Custom generators
 
 genNewType :: Gen BigTypeNewtype
 genNewType = BigTypeNewtype <$> genZoned

--- a/test/Test/Toml/BiCode/Property.hs
+++ b/test/Test/Toml/BiCode/Property.hs
@@ -1,0 +1,191 @@
+module Test.Toml.BiCode.Property where
+
+import Control.Applicative ((<|>))
+import Control.Category ((>>>))
+import Data.ByteString (ByteString)
+import Data.Hashable (Hashable)
+import Data.HashSet (HashSet)
+import Data.IntSet (IntSet)
+import Data.List.NonEmpty (NonEmpty)
+import Data.Set (Set)
+import Data.Text (Text)
+import Data.Time (Day, LocalTime, TimeOfDay, ZonedTime, zonedTimeToUTC)
+import GHC.Exts (fromList)
+import Hedgehog (Gen, forAll, tripping)
+import Numeric.Natural (Natural)
+
+import Toml (TomlBiMap, TomlCodec, (.=))
+import Toml.Bi.Code (decode, encode)
+
+import Test.Toml.Gen (PropertyTest, genBool, genDay, genDouble, genHours, genInteger, genLocal,
+                      genText, genZoned, prop)
+
+import qualified Data.ByteString.Lazy as L
+import qualified Data.Text as T
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import qualified Toml
+
+test_encodeDecodeProp :: PropertyTest
+test_encodeDecodeProp = prop "decode . encode == id" $ do
+    bigType <- forAll genBigType
+    tripping bigType (encode bigTypeCodec) (decode bigTypeCodec)
+
+data BigType = BigType
+    { btBool          :: Bool
+    , btInteger       :: Integer
+    , btNatural       :: Natural
+    , btInt           :: Int
+    , btWord          :: Word
+    , btDouble        :: Double
+    , btFloat         :: Float
+    , btText          :: Text
+    , btString        :: String
+    , btBS            :: ByteString
+    , btLazyBS        :: L.ByteString
+    , btLocalTime     :: LocalTime
+    , btDay           :: Day
+    , btTimeOfDay     :: TimeOfDay
+    , btArray         :: [Int]
+    , btArraySet      :: Set Word
+    , btArrayIntSet   :: IntSet
+    , btArrayHashSet  :: HashSet Double
+    , btArrayNonEmpty :: NonEmpty Text
+    , btNonEmpty      :: NonEmpty ByteString
+    , btList          :: [Bool]
+    , btNewtype       :: BigTypeNewtype
+    , btSum           :: BigTypeSum
+    , btRecord        :: BigTypeRecord
+    } deriving (Show, Eq)
+
+newtype BigTypeNewtype = BigTypeNewtype ZonedTime
+    deriving (Show)
+
+instance Eq BigTypeNewtype where
+    (BigTypeNewtype a) == (BigTypeNewtype b) = zonedTimeToUTC a == zonedTimeToUTC b
+
+data BigTypeSum = BigTypeSumA Integer | BigTypeSumB Text
+    deriving (Show, Eq)
+
+data BigTypeRecord = BigTypeRecord
+    { btrBoolSet     :: Set Bool
+    , btrNewtypeList :: [BigTypeSum]
+    } deriving (Show, Eq)
+
+bigTypeCodec :: TomlCodec BigType
+bigTypeCodec = BigType
+    <$> Toml.bool                            "bool"                .= btBool
+    <*> Toml.integer                         "integer"             .= btInteger
+    <*> Toml.natural                         "natural"             .= btNatural
+    <*> Toml.int                             "int"                 .= btInt
+    <*> Toml.word                            "word"                .= btWord
+    <*> Toml.double                          "double"              .= btDouble
+    <*> Toml.float                           "float"               .= btFloat
+    <*> Toml.text                            "text"                .= btText
+    <*> Toml.string                          "string"              .= btString
+    <*> Toml.byteString                      "bs"                  .= btBS
+    <*> Toml.lazyByteString                  "lbs"                 .= btLazyBS
+    <*> Toml.localTime                       "localTime"           .= btLocalTime
+    <*> Toml.day                             "day"                 .= btDay
+    <*> Toml.timeOfDay                       "timeOfDay"           .= btTimeOfDay
+    <*> Toml.arrayOf Toml._Int               "arrayOfInt"          .= btArray
+    <*> Toml.arraySetOf Toml._Word           "arraySetOfWord"      .= btArraySet
+    <*> Toml.arrayIntSet                     "arrayIntSet"         .= btArrayIntSet
+    <*> Toml.arrayHashSetOf Toml._Double     "arrayHashSetDouble"  .= btArrayHashSet
+    <*> Toml.arrayNonEmptyOf Toml._Text      "arrayNonEmptyOfText" .= btArrayNonEmpty
+    <*> Toml.nonEmpty (Toml.byteString "bs") "nonEmptyBS"          .= btNonEmpty
+    <*> Toml.list (Toml.bool "bool")         "listBool"            .= btList
+    <*> Toml.diwrap (Toml.zonedTime "nt.zonedTime")                .= btNewtype
+    <*> bigTypeSumCodec                                            .= btSum
+    <*> Toml.table bigTypeRecordCodec        "table-record"        .= btRecord
+
+_BigTypeSumA :: TomlBiMap BigTypeSum Integer
+_BigTypeSumA = Toml.prism BigTypeSumA $ \case
+    BigTypeSumA i -> Right i
+    other   -> Toml.wrongConstructor "BigTypeSumA" other
+
+_BigTypeSumB :: TomlBiMap BigTypeSum Text
+_BigTypeSumB = Toml.prism BigTypeSumB $ \case
+    BigTypeSumB n -> Right n
+    other    -> Toml.wrongConstructor "BigTypeSumB" other
+
+bigTypeSumCodec :: TomlCodec BigTypeSum
+bigTypeSumCodec =
+        Toml.match (_BigTypeSumA >>> Toml._Integer) "integer"
+    <|> Toml.match (_BigTypeSumB >>> Toml._Text)    "text"
+
+bigTypeRecordCodec :: TomlCodec BigTypeRecord
+bigTypeRecordCodec = BigTypeRecord
+    <$> Toml.arraySetOf Toml._Bool "rboolSet" .= btrBoolSet
+    <*> Toml.list bigTypeSumCodec "rnewtype" .= btrNewtypeList
+
+----------------------------------------------------------------------------
+-- Generator
+----------------------------------------------------------------------------
+
+genBigType :: Gen BigType
+genBigType = do
+    btBool          <- genBool
+    btInteger       <- genInteger
+    btNatural       <- fromInteger <$> genInteger
+    btInt           <- genInt
+    btWord          <- genWord
+    btDouble        <- genDouble
+    btFloat         <- genFloat
+    btText          <- genText
+    btString        <- genString
+    btBS            <- genBS
+    btLazyBS        <- L.fromStrict <$> genBS
+    btLocalTime     <- genLocal
+    btDay           <- genDay
+    btTimeOfDay     <- genHours
+    btArray         <- Gen.list (Range.constant 0 256) genInt
+    btArraySet      <- Gen.set (Range.constant 0 256) genWord
+    btArrayIntSet   <- genIntSet
+    btArrayHashSet  <- genHashSet genDouble
+    btArrayNonEmpty <- genNonEmpty genText
+    btNonEmpty      <- genNonEmpty genBS
+    btList          <- Gen.list (Range.constant 0 256) genBool
+    btNewtype       <- genNewType
+    btSum           <- genSum
+    btRecord        <- genRec
+    pure BigType {..}
+
+genInt :: Gen Int
+genInt = Gen.int (Range.constant 0 256)
+
+genWord :: Gen Word
+genWord = Gen.word (Range.constant 1 256)
+
+genFloat :: Gen Float
+genFloat = Gen.float (Range.constant 0 256)
+
+genString :: Gen String
+genString = T.unpack <$> genText
+
+genBS :: Gen ByteString
+genBS = Gen.bytes (Range.constant 0 256)
+
+genIntSet :: Gen IntSet
+genIntSet = fromList <$> Gen.list (Range.constant 0 256) genInt
+
+genHashSet :: (Eq a, Hashable a) => Gen a -> Gen (HashSet a)
+genHashSet genA = fromList <$> Gen.list (Range.constant 0 256) genA
+
+genNonEmpty :: Gen a -> Gen (NonEmpty a)
+genNonEmpty = Gen.nonEmpty (Range.constant 1 256)
+
+genNewType :: Gen BigTypeNewtype
+genNewType = BigTypeNewtype <$> genZoned
+
+genSum :: Gen BigTypeSum
+genSum = Gen.choice
+    [ BigTypeSumA <$> genInteger
+    , BigTypeSumB <$> genText
+    ]
+
+genRec :: Gen BigTypeRecord
+genRec = do
+    btrBoolSet <- Gen.set (Range.constant 0 256) genBool
+    btrNewtypeList <- Gen.list (Range.constant 0 256) genSum
+    pure BigTypeRecord{..}

--- a/test/Test/Toml/BiCode/Property.hs
+++ b/test/Test/Toml/BiCode/Property.hs
@@ -37,8 +37,8 @@ data BigType = BigType
     , btNatural       :: Natural
     , btInt           :: Int
     , btWord          :: Word
-    , btDouble        :: Double
-    , btFloat         :: Float
+    , btDouble        :: Batman Double
+    , btFloat         :: Batman Float
     , btText          :: Text
     , btString        :: String
     , btBS            :: ByteString
@@ -49,7 +49,7 @@ data BigType = BigType
     , btArray         :: [Int]
     , btArraySet      :: Set Word
     , btArrayIntSet   :: IntSet
-    , btArrayHashSet  :: HashSet Double
+    , btArrayHashSet  :: HashSet Natural
     , btArrayNonEmpty :: NonEmpty Text
     , btNonEmpty      :: NonEmpty ByteString
     , btList          :: [Bool]
@@ -57,6 +57,16 @@ data BigType = BigType
     , btSum           :: BigTypeSum
     , btRecord        :: BigTypeRecord
     } deriving (Show, Eq)
+
+-- | Wrapper over 'Double' and 'Float' to be equal on @NaN@ values.
+newtype Batman a = Batman a
+    deriving (Show)
+
+instance RealFloat a => Eq (Batman a) where
+    Batman a == Batman b =
+        if isNaN a
+            then isNaN b
+            else a == b
 
 newtype BigTypeNewtype = BigTypeNewtype ZonedTime
     deriving (Show)
@@ -79,8 +89,8 @@ bigTypeCodec = BigType
     <*> Toml.natural                         "natural"             .= btNatural
     <*> Toml.int                             "int"                 .= btInt
     <*> Toml.word                            "word"                .= btWord
-    <*> Toml.double                          "double"              .= btDouble
-    <*> Toml.float                           "float"               .= btFloat
+    <*> Toml.diwrap (Toml.double "double")                         .= btDouble
+    <*> Toml.diwrap (Toml.float "float")                           .= btFloat
     <*> Toml.text                            "text"                .= btText
     <*> Toml.string                          "string"              .= btString
     <*> Toml.byteString                      "bs"                  .= btBS
@@ -91,7 +101,7 @@ bigTypeCodec = BigType
     <*> Toml.arrayOf Toml._Int               "arrayOfInt"          .= btArray
     <*> Toml.arraySetOf Toml._Word           "arraySetOfWord"      .= btArraySet
     <*> Toml.arrayIntSet                     "arrayIntSet"         .= btArrayIntSet
-    <*> Toml.arrayHashSetOf Toml._Double     "arrayHashSetDouble"  .= btArrayHashSet
+    <*> Toml.arrayHashSetOf Toml._Natural    "arrayHashSetDouble"  .= btArrayHashSet
     <*> Toml.arrayNonEmptyOf Toml._Text      "arrayNonEmptyOfText" .= btArrayNonEmpty
     <*> Toml.nonEmpty (Toml.byteString "bs") "nonEmptyBS"          .= btNonEmpty
     <*> Toml.list (Toml.bool "bool")         "listBool"            .= btList
@@ -111,8 +121,8 @@ _BigTypeSumB = Toml.prism BigTypeSumB $ \case
 
 bigTypeSumCodec :: TomlCodec BigTypeSum
 bigTypeSumCodec =
-        Toml.match (_BigTypeSumA >>> Toml._Integer) "integer"
-    <|> Toml.match (_BigTypeSumB >>> Toml._Text)    "text"
+        Toml.match (_BigTypeSumA >>> Toml._Integer) "sum.integer"
+    <|> Toml.match (_BigTypeSumB >>> Toml._Text)    "sum.text"
 
 bigTypeRecordCodec :: TomlCodec BigTypeRecord
 bigTypeRecordCodec = BigTypeRecord
@@ -127,11 +137,11 @@ genBigType :: Gen BigType
 genBigType = do
     btBool          <- genBool
     btInteger       <- genInteger
-    btNatural       <- fromInteger <$> genInteger
+    btNatural       <- genNatural
     btInt           <- genInt
     btWord          <- genWord
-    btDouble        <- genDouble
-    btFloat         <- genFloat
+    btDouble        <- Batman <$> genDouble
+    btFloat         <- Batman <$> genFloat
     btText          <- genText
     btString        <- genString
     btBS            <- genBS
@@ -139,23 +149,26 @@ genBigType = do
     btLocalTime     <- genLocal
     btDay           <- genDay
     btTimeOfDay     <- genHours
-    btArray         <- Gen.list (Range.constant 0 256) genInt
-    btArraySet      <- Gen.set (Range.constant 0 256) genWord
+    btArray         <- Gen.list (Range.constant 0 5) genInt
+    btArraySet      <- Gen.set (Range.constant 0 5) genWord
     btArrayIntSet   <- genIntSet
-    btArrayHashSet  <- genHashSet genDouble
+    btArrayHashSet  <- genHashSet genNatural
     btArrayNonEmpty <- genNonEmpty genText
     btNonEmpty      <- genNonEmpty genBS
-    btList          <- Gen.list (Range.constant 0 256) genBool
+    btList          <- Gen.list (Range.constant 0 5) genBool
     btNewtype       <- genNewType
     btSum           <- genSum
     btRecord        <- genRec
     pure BigType {..}
 
 genInt :: Gen Int
-genInt = Gen.int (Range.constant 0 256)
+genInt = Gen.int $ Range.constant 0 256
+
+genNatural :: Gen Natural
+genNatural = fromIntegral <$> Gen.int (Range.constant 1 256)
 
 genWord :: Gen Word
-genWord = Gen.word (Range.constant 1 256)
+genWord = Gen.word $ Range.constant 0 256
 
 genFloat :: Gen Float
 genFloat = Gen.float (Range.constant 0 256)
@@ -164,16 +177,16 @@ genString :: Gen String
 genString = T.unpack <$> genText
 
 genBS :: Gen ByteString
-genBS = Gen.bytes (Range.constant 0 256)
+genBS = Gen.utf8 (Range.constant 0 20) Gen.alphaNum
 
 genIntSet :: Gen IntSet
-genIntSet = fromList <$> Gen.list (Range.constant 0 256) genInt
+genIntSet = fromList <$> Gen.list (Range.constant 0 5) genInt
 
 genHashSet :: (Eq a, Hashable a) => Gen a -> Gen (HashSet a)
-genHashSet genA = fromList <$> Gen.list (Range.constant 0 256) genA
+genHashSet genA = fromList <$> Gen.list (Range.constant 0 5) genA
 
 genNonEmpty :: Gen a -> Gen (NonEmpty a)
-genNonEmpty = Gen.nonEmpty (Range.constant 1 256)
+genNonEmpty = Gen.nonEmpty (Range.constant 1 5)
 
 genNewType :: Gen BigTypeNewtype
 genNewType = BigTypeNewtype <$> genZoned
@@ -186,6 +199,6 @@ genSum = Gen.choice
 
 genRec :: Gen BigTypeRecord
 genRec = do
-    btrBoolSet <- Gen.set (Range.constant 0 256) genBool
-    btrNewtypeList <- Gen.list (Range.constant 0 256) genSum
+    btrBoolSet <- fromList <$> Gen.list (Range.constant 0 5) genBool
+    btrNewtypeList <- Gen.list (Range.constant 0 5) genSum
     pure BigTypeRecord{..}

--- a/test/Test/Toml/BiMap/Property.hs
+++ b/test/Test/Toml/BiMap/Property.hs
@@ -2,51 +2,17 @@
 
 module Test.Toml.BiMap.Property where
 
-import Hedgehog (Gen, PropertyT, Range, assert, forAll, tripping, (===))
+import Hedgehog (Gen, PropertyT, assert, forAll, tripping, (===))
 
-import Data.Time (Day, ZonedTime (..))
-import GHC.Natural (Natural)
+import Data.Time (ZonedTime (..))
 import Test.Tasty (testGroup)
 import Test.Toml.Gen (PropertyTest, prop)
 import Toml.Bi.Map (BiMap (..), TomlBiMap)
 
-import qualified Data.ByteString.Builder as BB
-import qualified Data.ByteString.Lazy as BL
-import qualified Data.HashSet as HS
-import qualified Data.IntSet as IS
-import qualified Data.List.NonEmpty as NE
-import qualified Data.Text.Lazy as L
 import qualified Hedgehog.Gen as Gen
-import qualified Hedgehog.Range as Range
 import qualified Test.Toml.Gen as G
 import qualified Toml.Bi.Map as B
 
--- Custom generators
-
-range100 :: Range Int
-range100 = Range.constant 0 100
-
-genNatural :: Gen Natural
-genNatural = fromIntegral <$> Gen.word Range.constantBounded
-
-genInSet :: Gen IS.IntSet
-genInSet = IS.fromList <$> Gen.list range100 (Gen.int Range.constantBounded)
-
-genHashSet :: Gen (HS.HashSet Integer)
-genHashSet = HS.fromList <$> Gen.list range100 G.genInteger
-
-genNEDays :: Gen (NE.NonEmpty Day)
-genNEDays = Gen.nonEmpty (Range.constant 1 100) G.genDay
-
-genLByteString :: Gen BL.ByteString
-genLByteString =
-  BB.toLazyByteString . BB.byteString <$>
-  Gen.utf8 range100 Gen.unicodeAll
-
-genLText :: Gen L.Text
-genLText = L.fromStrict <$> G.genText
-
--- Tests
 
 testBiMap
     :: (Monad m, Show a, Show b, Eq a)
@@ -70,30 +36,26 @@ test_BiMaps :: PropertyTest
 test_BiMaps = pure $ testGroup "BiMap roundtrip tests" $ concat
     [ prop "Bool" (testBiMap B._Bool G.genBool)
     , prop "Integer" (testBiMap B._Integer G.genInteger)
-    , prop "Natural" (testBiMap B._Natural genNatural)
-    , prop "Int" (testBiMap B._Int (Gen.int Range.constantBounded))
-    , prop "Word" (testBiMap B._Word (Gen.word Range.constantBounded))
+    , prop "Natural" (testBiMap B._Natural G.genNatural)
+    , prop "Int" (testBiMap B._Int G.genInt)
+    , prop "Word" (testBiMap B._Word G.genWord)
     , prop "Double" testDouble
-    , prop "Float"
-        (testBiMap B._Float (Gen.float $ Range.constant (-10000.0) 10000.0))
+    , prop "Float" (testBiMap B._Float G.genFloat)
     , prop "Text" (testBiMap B._Text G.genText)
-    , prop "LazyText" (testBiMap B._LText genLText)
-    , prop "String"
-        (testBiMap B._String (Gen.string range100 Gen.unicode))
+    , prop "LazyText" (testBiMap B._LText G.genLText)
+    , prop "String" (testBiMap B._String G.genString)
     , prop "Read (Integer)" (testBiMap B._Read G.genInteger)
-    , prop "ByteString"
-        (testBiMap B._ByteString (Gen.utf8 range100 Gen.unicodeAll))
-    , prop "Lazy ByteString" (testBiMap B._LByteString genLByteString)
+    , prop "ByteString" (testBiMap B._ByteString G.genByteString)
+    , prop "Lazy ByteString" (testBiMap B._LByteString G.genLByteString)
     , prop "ZonedTime" (testBiMap B._ZonedTime G.genZoned)
     , prop "LocalTime" (testBiMap B._LocalTime G.genLocal)
     , prop "TimeOfDay" (testBiMap B._TimeOfDay G.genHours)
     , prop "Day" (testBiMap B._Day G.genDay)
-    , prop "IntSet" (testBiMap B._IntSet genInSet)
-    , prop "Array (Day)"
-        (testBiMap (B._Array B._Day) (Gen.list range100 G.genDay))
-    , prop "Set (Day)" (testBiMap (B._Set B._Day) (Gen.set range100 G.genDay))
-    , prop "NonEmpty (Day)" (testBiMap (B._NonEmpty B._Day) genNEDays)
-    , prop "HashSet (Integer)" (testBiMap (B._HashSet B._Integer) genHashSet)
+    , prop "IntSet" (testBiMap B._IntSet G.genIntSet)
+    , prop "Array (Day)" (testBiMap (B._Array B._Day) (G.genList G.genDay))
+    , prop "Set (Day)" (testBiMap (B._Set B._Day) (Gen.set G.range100 G.genDay))
+    , prop "NonEmpty (Day)" (testBiMap (B._NonEmpty B._Day) (G.genNonEmpty G.genDay))
+    , prop "HashSet (Integer)" (testBiMap (B._HashSet B._Integer) (G.genHashSet G.genInteger))
     ]
 
 -- Orphan instances

--- a/test/Test/Toml/Gen.hs
+++ b/test/Test/Toml/Gen.hs
@@ -11,40 +11,72 @@ module Test.Toml.Gen
        , prop
 
          -- * Generators
-       , genVal
-       , genKey
-       , genPrefixMap
-       , genToml
-       , genBool
+         -- ** Primitive
+       , genInt
        , genInteger
+       , genDouble
+       , genWord
+       , genNatural
+       , genFloat
+
+       , genList
+       , genNonEmpty
+       , genHashSet
+       , genIntSet
+
+       , genBool
+
+       , genText
+       , genString
+       , genByteString
+       , genLByteString
+       , genLText
+
+         -- ** Dates
        , genDay
        , genHours
        , genLocal
        , genZoned
-       , genDouble
-       , genText
+
+         -- ** @TOML@ specific
+       , genVal
+       , genKey
+       , genPrefixMap
+       , genToml
+
+         -- ** Other
+       , range100
        ) where
 
 import Control.Applicative (liftA2)
 import Control.Monad (forM, replicateM)
+import Data.ByteString (ByteString)
 import Data.Fixed (Fixed (..))
+import Data.Hashable (Hashable)
+import Data.HashSet (HashSet)
+import Data.IntSet (IntSet)
+import Data.List.NonEmpty (NonEmpty)
 import Data.Text (Text)
 import Data.Time (Day, LocalTime (..), TimeOfDay (..), ZonedTime (..), fromGregorian,
                   minutesToTimeZone)
+import GHC.Exts (fromList)
 import GHC.Stack (HasCallStack)
-import Hedgehog (MonadGen, PropertyT, property)
+import Hedgehog (Gen, MonadGen, PropertyT, Range, property)
+import Numeric.Natural (Natural)
 import Test.Tasty (TestName, TestTree)
 import Test.Tasty.Hedgehog (testProperty)
 
 import Toml.Bi.Map (toMArray)
-import Toml.PrefixTree (pattern (:||), Key (..), Piece (..), PrefixMap, PrefixTree (..), fromList)
+import Toml.PrefixTree (pattern (:||), Key (..), Piece (..), PrefixMap, PrefixTree (..))
 import Toml.Type (AnyValue (..), TOML (..), TValue (..), Value (..))
 
-import qualified Data.HashMap.Strict as HashMap
+import qualified Data.ByteString.Lazy as LB
 import qualified Data.List.NonEmpty as NE
 import qualified Data.Text as Text
+import qualified Data.Text.Lazy as L
 import qualified Hedgehog.Gen as Gen
 import qualified Hedgehog.Range as Range
+import qualified Toml.PrefixTree as Toml (fromList)
 
 ----------------------------------------------------------------------------
 -- Property test creator
@@ -58,6 +90,8 @@ prop testName = pure . testProperty testName . property
 ----------------------------------------------------------------------------
 -- Common generators
 ----------------------------------------------------------------------------
+
+-- @TOML@ specific
 
 type V = Int
 
@@ -94,7 +128,7 @@ genPrefixMap = do
         tree <- genPrefixTree key
         pure (piece, tree)
 
-    pure $ HashMap.fromList kvps
+    pure $ fromList kvps
 
 genPrefixTree :: forall m . MonadGen m => Key -> m (PrefixTree V)
 genPrefixTree key = Gen.recursive
@@ -111,7 +145,7 @@ genPrefixTree key = Gen.recursive
         pure $ Branch key prefVal prefMap
 
 makeToml :: [(Key, AnyValue)] -> TOML
-makeToml kv = TOML (HashMap.fromList kv) mempty mempty
+makeToml kv = TOML (fromList kv) mempty mempty
 
 genToml :: MonadGen m => m TOML
 genToml = Gen.recursive
@@ -119,15 +153,17 @@ genToml = Gen.recursive
             [ makeToml <$> genKeyAnyValueList ]
             [ TOML <$> keyValues <*> tables <*> arrays ]
   where
-    keyValues = HashMap.fromList <$> genKeyAnyValueList
-    tables = fmap fromList
+    keyValues = fromList <$> genKeyAnyValueList
+    tables = fmap Toml.fromList
              $ Gen.list (Range.linear 0 5)
              $ (,) <$> genKey <*> genToml
-    arrays = fmap HashMap.fromList $
+    arrays = fmap fromList $
              Gen.list (Range.linear 0 5) $ do
                key <- genKey
                arr <- Gen.list (Range.linear 1 5) genToml
                return (key, NE.fromList arr)
+
+-- Date generators
 
 genDay :: MonadGen m => m Day
 genDay = do
@@ -155,11 +191,16 @@ genZoned = do
     let zTime = minutesToTimeZone zMin
     pure $ ZonedTime local zTime
 
-genBool :: MonadGen m => m Bool
-genBool = Gen.bool
+-- Primitive generators
+
+range100 :: Range Int
+range100 = Range.constant 0 100
+
+genInt :: MonadGen m => m Int
+genInt = Gen.int Range.constantBounded
 
 genInteger :: MonadGen m => m Integer
-genInteger = toInteger <$> Gen.int (Range.constantBounded @Int)
+genInteger = toInteger <$> genInt
 
 genDouble :: MonadGen m => m Double
 genDouble = Gen.frequency
@@ -168,6 +209,30 @@ genDouble = Gen.frequency
     , (1, Gen.constant $ -1/0)
     , (1, Gen.constant $ 0/0)
     ]
+
+genWord :: MonadGen m => m Word
+genWord = Gen.word Range.constantBounded
+
+genNatural :: MonadGen m => m Natural
+genNatural = fromIntegral <$> Gen.word Range.constantBounded
+
+genFloat :: MonadGen m => m Float
+genFloat = Gen.float (Range.constant (-10000.0) 10000.0)
+
+genHashSet :: (Eq a, Hashable a) => Gen a -> Gen (HashSet a)
+genHashSet genA = fromList <$> genList genA
+
+genNonEmpty :: Gen a -> Gen (NonEmpty a)
+genNonEmpty = Gen.nonEmpty (Range.constant 1 5)
+
+genList :: Gen a -> Gen [a]
+genList = Gen.list range100
+
+genIntSet :: Gen IntSet
+genIntSet = fromList <$> genList genInt
+
+genBool :: MonadGen m => m Bool
+genBool = Gen.bool
 
 -- | Generatates control sympol.
 genEscapeSequence :: MonadGen m => m Text
@@ -207,6 +272,18 @@ genText =  fmap Text.concat $ Gen.list (Range.constant 0 256) $ Gen.choice
     , genUniHex4Color
     , genUniHex8Color
     ]
+
+genString :: Gen String
+genString = Text.unpack <$> genText
+
+genByteString :: Gen ByteString
+genByteString = Gen.utf8 range100 Gen.alphaNum
+
+genLByteString :: Gen LB.ByteString
+genLByteString = LB.fromStrict <$> genByteString
+
+genLText :: Gen L.Text
+genLText = L.fromStrict <$> genText
 
 -- | List of AnyValue generators.
 noneArrayList :: MonadGen m => [m AnyValue]

--- a/test/Test/Toml/Gen.hs
+++ b/test/Test/Toml/Gen.hs
@@ -214,7 +214,7 @@ genWord :: MonadGen m => m Word
 genWord = Gen.word Range.constantBounded
 
 genNatural :: MonadGen m => m Natural
-genNatural = fromIntegral <$> Gen.word Range.constantBounded
+genNatural = fromIntegral <$> genWord
 
 genFloat :: MonadGen m => m Float
 genFloat = Gen.float (Range.constant (-10000.0) 10000.0)

--- a/test/Test/Toml/Gen.hs
+++ b/test/Test/Toml/Gen.hs
@@ -64,7 +64,6 @@ type V = Int
 genVal :: MonadGen m => m V
 genVal = Gen.int (Range.constant 0 256)
 
--- TODO: Arrays and Date.
 -- | Generates random value of 'AnyValue' type.
 genAnyValue :: MonadGen m => m AnyValue
 genAnyValue = Gen.choice $
@@ -164,10 +163,10 @@ genInteger = toInteger <$> Gen.int (Range.constantBounded @Int)
 
 genDouble :: MonadGen m => m Double
 genDouble = Gen.frequency
-    [ (50, Gen.double $ Range.constant @Double (-1000000.0) 1000000.0)
-    , (5, Gen.constant $ 1/0)
-    , (5, Gen.constant $ -1/0)
-    , (5, Gen.constant $ 0/0)
+    [ (10, Gen.double $ Range.constant @Double (-1000000.0) 1000000.0)
+    , (1, Gen.constant $ 1/0)
+    , (1, Gen.constant $ -1/0)
+    , (1, Gen.constant $ 0/0)
     ]
 
 -- | Generatates control sympol.

--- a/tomland.cabal
+++ b/tomland.cabal
@@ -127,6 +127,7 @@ test-suite tomland-test
   main-is:             Spec.hs
 
   other-modules:       Test.Toml.BiMap.Property
+                       Test.Toml.BiCode.Property
                        Test.Toml.Gen
                        Test.Toml.Property
                        Test.Toml.Parsing.Property
@@ -140,6 +141,7 @@ test-suite tomland-test
   build-depends:       base
                      , bytestring ^>= 0.10
                      , containers >= 0.5.7 && < 0.7
+                     , hashable
                      , hedgehog ^>= 0.6
                      , hspec-megaparsec
                      , megaparsec

